src/PaymentExporter.php

```diff
@@ -7,29 +7,86 @@
 
+  protected $submissionExporter;
+
+  /**
+   * Create a new payment exporter.
+   */
+  public function __construct(SubmissionExporter $submission_exporter) {
+    $this->submissionExporter = $submission_exporter;
+  }
+
   /**
-   * Export a payment object to something we want to transmit in logCRM events.
+   * Generate the event data for a payment object.
    *
    * @param \Payment $payment
-   *   The pament to export.
+   *   The payment to export.
    *
    * @return array
-   *   An associative array of data for this payment.
+   *   An associative array with the payment’s data.
    */
-  public function toJson(\Payment $payment) {
-    $data['pid'] = $payment->pid;
+  public function paymentData(\Payment $payment) {
     $status = $payment->getStatus();
-    $data['currency_code'] = $payment->currency_code;
-    $data['total_amount'] = $payment->totalAmount(TRUE);
-    $data['status'] = $status->status;
-    $data['method_specific'] = $payment->method->title_specific;
-    $data['method_generic'] = $payment->method->title_generic;
     $controller = $payment->method->controller;
-    $data['controller'] = $controller->name;
+    $data = [
+      'pid' => $payment->pid,
+      'currency_code' => $payment->currency_code,
+      'total_amount' => $payment->totalAmount(TRUE),
+      'status' => $status->status,
+      'method_specific' => $payment->method->title_specific,
+      'method_generic' => $payment->method->title_generic,
+      'controller' => $controller->name,
+    ];
     if (webform_paymethod_select_implements_data_interface($controller)) {
       $data['payment_data'] = $controller->webformData($payment);
     }
+
+    // Let other modules alter the data.
+    // Deprecated: Use campaignion_logcrm_event_data_alter() instead.
+    drupal_alter('campaignion_logcrm_payment_event_data', $data, $payment);
     return $data;
   }
 
+  /**
+   * Create a payment_success event from a payment object.
+   *
+   * @param \Payment $payment
+   *   The payment to export.
+   *
+   * @return Event
+   *   A webhook event with data for the payment.
+   */
+  public function createSuccessEvent(\Payment $payment) {
+    $data = $this->paymentData($payment);
+    $submission = $payment->contextObj->getSubmission();
+    $data['uuid'] = $submission->uuid;
+    $data['action'] = $this->submissionExporter->actionData($submission);
+
+    $context['payment'] = $payment;
+    $context['submission'] = $submission;
+    return Event::fromData('payment_success', $payment->getStatus()->created, $data, $context);
+  }
+
+  /**
+   * Create a payment_status_change event.
+   *
+   * @param \Payment $payment
+   *   The payment for which to report the status change.
+   * @param \PaymentStatusItem $prev_status_item
+   *   The previous status item.
+   *
+   * @return Event
+   *   A logCRM event with data for the payment status change.
+   */
+  public function createStatusChangeEvent(\Payment $payment, \PaymentStatusItem $prev_status_item) {
```

Comment:
The only difference between "success" events and "status change" events is that "status change" includes the previous status while "success" includes the action data?

According to https://github.com/moreonion/impact-stack-docs/pull/3 the "success" event is about to be deprecated, maybe this could be reflected in the code? Is the plan to remove all queue item creations with `$exporter->createSuccessEvent($payment)`?

Reply:
So far the status change events also don’t include the `action` object. The biggest difference though is that there might be more than one such status item per payment (as opposed to just one).

Reply:
Thank you for adding a deprecation notice! It’s missing a few "d", but still clear enough :wink: